### PR TITLE
Story/q20 19

### PR DESCRIPTION
**Pull request by Pooja444**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed)
 
https://diging.atlassian.net/browse/Q20-19
On the collection page, it should be possible to browse all networks in collection - 
basically a paginated list of all networks ordered by submission date. When the user clicks on a network it would be visualized (similar to the old quadriga).
  
## Anything else the reviewer needs to know?

This story is dependent on story Q20-3, Q20-7, Q20-12, Q20-18 and since all of these stories are yet to be merged, this PR may show a lot of file changes that are actually part of the above stories. This will go away once all of these are merged

### CODE REVIEW

**Comment by Pooja444** [comment]
As this PR is dependent on Q20-18, will re-open it once Q20-18 is merged

**Comment by jdamerow** [comment]
Resolve conflicts please

**Comment by jdamerow** [comment]
Resolve conflicts please

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+import org.hibernate.annotations.LazyCollection;
+import org.hibernate.annotations.LazyCollectionOption;
+
+/**
+ * This class represents the 'conceptpower_concept_cache' table present in the
+ * database that stores cached concept data
+ * 
+ * @author poojakulkarni
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java]: this class looks like a lot came from the concept class in the previous quadriga project, so author parameter should list the Digital Innovation Group since there is no author in the javadoc of the original class.

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            if (response != null) {
+                return response.getBody();
+            } else {
+                logger.error("No response returned from ConceptPower for URI: " + conceptURI);
+                if (conceptPowerURL == null || conceptPowerURL.equals("")) {
+                    logger.error("ConceptPowerURL was found to be blank or null");
+                }
+            }
+        } catch (RestClientException e) {
+            logger.error(e.getMessage());
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java]: the whole exception (with stacktrace needs to be logged)

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    private ConceptPowerConnectorService conceptPowerConnectorService;
+
+    private Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public ConceptCache getConceptByUri(String uri) {
+
+        ConceptCache conceptCache = conceptCacheService.getConceptByUri(uri);
+
+        if (conceptCache == null || ChronoUnit.DAYS.between(conceptCache.getLastUpdated(), LocalDateTime.now()) >= 2) {
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: the time how long until a concept is updated again needs to be configurable. Also,  let's go with hours rather than days.

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+     *                          was generated
+     */
+    private ConceptCache saveConceptCacheFromConceptPowerReply(ConceptCache conceptCacheOld, ConceptPowerReply conceptPowerReply, String uri) {
+        ConceptCache conceptCache = mapConceptPowerReplyToConceptCache(conceptPowerReply);
+        
+        // Before returning, we need to check if we've updated ConceptCache or not
+        // If no diff was found, conceptCache won't be updated and 'lastUpdated' would stay the same
+        boolean updated = false;
+        
+        updated = updateConceptCache(conceptCache, conceptCacheOld, uri);
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: merge with line 81

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/data/EventGraphRepository.java [hunk 5]
```diff
@@ -5,11 +5,19 @@
...
 import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.data.mongodb.repository.Query;
 
 import edu.asu.diging.quadriga.core.model.EventGraph;
 
 public interface EventGraphRepository extends MongoRepository<EventGraph, ObjectId> {
+    
+    public Optional<List<EventGraph>> findByCollectionId(ObjectId collectionId);
+    
+    @Query(value = "{'context.sourceUri': ?0}")
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/data/EventGraphRepository.java]: is this needed? does autogenerating the query not work here?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/model/EventGraph.java [hunk 6]
```diff
@@ -13,6 +13,9 @@ public class EventGraph {
     private DefaultMapping defaultMapping;
     private OffsetDateTime creationTime;
     private ObjectId collectionId;
+    private String appName;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/model/EventGraph.java]: what's the difference between "appName" and "submittingApp"?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/service/impl/MappedTripleGroupServiceImpl.java [hunk 7]
```diff
@@ -145,6 +150,9 @@ public MappedTripleGroup get(String collectionId, MappedTripleType mappedTripleT
         if (mappedTripleGroup == null) {
             mappedTripleGroup = add(collectionId, mappedTripleType);
         }
+        if (mappedTripleGroup == null) {
+            logger.error("Couldn't find or persist a new MappedTripleGroup entry for collectionId: " + collectionId);
+        }
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/service/impl/MappedTripleGroupServiceImpl.java]: I would move this if statement inside the above if statement. This is confusing and really only can happen if line 151 doesn't return a collection, so no need for it to be outside the if.

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 8]
```diff
@@ -36,43 +45,71 @@ public class DisplayCollectionController {
...
-       
-        
+
+        model.addAttribute("collection", collection);
+
+        // Determine page number and size for network pagination
+        int page = 0;
+        int size = 10;
+
+        if (request.getParameter("page") != null && !request.getParameter("page").isEmpty()) {
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: request parameter should be injected into the method via spring (`@RequestParam`).

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 9]
```diff
@@ -36,43 +45,71 @@ public class DisplayCollectionController {
...
+            page = Integer.parseInt(request.getParameter("page")) - 1;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: you probably want to catch the runtime exception `parseInt` throws if the value is not actually. a number.

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 10]
```diff
@@ -36,43 +45,71 @@ public class DisplayCollectionController {
...
+            page = page < 0 ? 0 : page;
+        }
+
+        if (request.getParameter("size") != null && !request.getParameter("size").isEmpty()) {
+            size = Integer.parseInt(request.getParameter("size"));
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: same as above for these two llines

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 11]
```diff
@@ -36,43 +45,71 @@ public class DisplayCollectionController {
...
-        
+        // Get all EventGraphs for this collection
+        List<EventGraph> eventGraphsList = eventGraphService.findAllEventGraphsByCollectionId(collection.getId());
+
+
         long numberOfSubmittedNetworks = eventGraphService.getNumberOfSubmittedNetworks(collection.getId());
-       
+
+
+        model.addAttribute("networks", eventGraphsList.subList(page * size, Math.min(eventGraphsList.size(), page * size + size)));
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: pagination should happen on db level not here.

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/NetworkController.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        if (!isSourceUriValid(sourceUri)) {
+            logger.error("Invalid sourceUri: " + sourceUri);
+            return "error404Page";
+        }
+
+        // List<EventGraph> eventGraphs = eventGraphService.findEventGraphsBySourceURI(sourceURI);
+        List<EventGraph> eventGraphs = eventGraphService.findAllEventGraphsByCollectionId(new ObjectId(collectionId));
+        
+        if (eventGraphs == null) {
+            logger.error("No network found for sourceUri: " + sourceUri);
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/NetworkController.java]: I'm confused, the event graphs are retrieved by collection id, arent' they? What has the source id to do with that?

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/model/GraphNodeType.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package edu.asu.diging.quadriga.web.model;
+
+public enum GraphNodeType {
+    
+    PREDICATE(0),
+    SUBJECT(1),
+    OBJECT(1);
+    
+    private int groupId;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/model/GraphNodeType.java:12]: what is `groupId`?

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/service/impl/GraphCreationServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        graphElements.setEdges(wrapInGraphElements(graphEdges));
+
+        return graphElements;
+    }
+
+    @Override
+    public String createNodesAndEdges(RelationEvent event, List<GraphData> graphNodes, List<GraphData> graphEdges,
+            Map<String, GraphNodeData> uniqueNodes, String eventGraphId) {
+        Relation relation = event.getRelation();
+        String predicateNodeId = null, subjectNodeId = null, objectNodeId = null;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/service/impl/GraphCreationServiceImpl.java]: this doesn't follow our code style; 
[https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations](https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations)

File: quadriga/src/main/resources/config.properties [hunk 15]
```diff
@@ -6,6 +6,9 @@ db.password=${db.password}
 neo4j.url=${neo4j.url}
 neo4j.database=${neo4j.database}
 
+conceptpower_base_url=${conceptpower.base.url}
```
Comment (jdamerow) [inline quadriga/src/main/resources/config.properties:9]: does `conceptpower.base.url` already exist in the pom?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
also, is there a reason this is still a draft?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+ * @author Digital Innovation Group
+ *
+ */
+@Entity
+@Table(name = "conceptpower_concept_cache")
+public class ConceptCache implements Serializable, Comparable<ConceptCache> {
+
+    /**
+     * 
+     */
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java]: remove if there is no javadoc

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+    private static final long serialVersionUID = 6260562492095202271L;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java]: should start at 1

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptType.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+@Entity
+@Table(name = "conceptpower_concept_type_cache")
+public class ConceptType implements Serializable, Comparable<ConceptType> {
+
+    /**
+     * 
+     */
+    private static final long serialVersionUID = 4228339422371096758L;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptType.java]: same as before

File: quadriga/pom.xml [hunk 19]
```diff
@@ -44,6 +44,9 @@
         <citesphere.base.url></citesphere.base.url>
         <citesphere.client.id></citesphere.client.id>
         <citesphere.client.secret></citesphere.client.secret>
+        
+        <conceptpower.base.url></conceptpower.base.url>
+        <conceptpower.id.url></conceptpower.id.url>
```
Comment (jdamerow) [inline quadriga/pom.xml]: is this used?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+@Service
+@PropertySource({ "classpath:config.properties" })
+public class ConceptPowerConnectorServiceImpl implements ConceptPowerConnectorService {
+
+    @Value("${conceptpower_base_url}")
+    private String conceptPowerBaseURL;
+
+    @Value("${conceptpower_id_url}")
+    private String conceptPowerIdURL;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java]: that's the endpoint not the url, right?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+    private RestTemplate restTemplate;
+
+    private Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java]: logger should come first in the class

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+                logger.error("No response returned from ConceptPower for URI: " + conceptURI);
+                if (conceptPowerURL == null || conceptPowerURL.equals("")) {
+                    logger.error("ConceptPowerURL was found to be blank or null");
+                }
+            }
+        } catch (RestClientException e) {
+            logger.error("Could not get concept for URI: " + conceptURI, e);
+            if (conceptPowerURL == null || conceptPowerURL.equals("")) {
+                logger.error("ConceptPowerURL was found to be blank or null");
+            }
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java:59]: instead of this extra if, I would just add to the log message in line 63 something like "at URL ...".

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+    @Autowired
+    private ConceptTypeService conceptTypeService;
+
+    @Autowired
+    private ConceptPowerConnectorService conceptPowerConnectorService;
+    
+    @Value("${conceptCacheUpdateInterval}")
+    private Integer conceptCacheUpdateInterval;
+
+    private Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: should be first

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+            conceptCache = new ConceptCache();
+            conceptCache.setUri(conceptEntry.getConceptUri());
+            conceptCache.setConceptList(conceptEntry.getConceptList());
+            conceptCache.setDescription(conceptEntry.getDescription());
+            conceptCache.setPos(conceptEntry.getPos());
+            conceptCache.setDeleted(conceptEntry.getDeleted() == null ? false : conceptEntry.getDeleted());
+            conceptCache.setCreatorId(conceptEntry.getCreatorId());
+            conceptCache.setWord(conceptEntry.getLemma());
+
+            if(conceptEntry.getConceptUri() != null) {
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: what this if for? Conceptpower returns the id as a separate field, doesn't it?

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 25]
```diff
@@ -36,43 +49,72 @@ public class DisplayCollectionController {
...
+        int sizeLimit = 10;
+
+        if (page != null) {
+            pageNumber = page - 1;
+            pageNumber = pageNumber < 0 ? 0 : pageNumber;
+        }
+
+        if (size != null) {
+            sizeLimit = size;
+            sizeLimit = sizeLimit < 1 ? 10 : sizeLimit;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: why not changing `size`? is a new variable necessary?

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/model/GraphNodeType.java (hunk 13 above)
Comment (jdamerow) [reply quadriga/src/main/java/edu/asu/diging/quadriga/web/model/GraphNodeType.java:12]: ? neesd javadoc

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/service/impl/GraphCreationServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+import edu.asu.diging.quadriga.web.model.GraphNodeType;
+import edu.asu.diging.quadriga.web.service.GraphCreationService;
+
+@Service
+public class GraphCreationServiceImpl implements GraphCreationService {
+
+    @Autowired
+    private ConceptPowerService conceptPowerService;
+
+    private Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/service/impl/GraphCreationServiceImpl.java]: shoudl be first

File: quadriga/src/main/resources/config.properties [hunk 27]
```diff
@@ -6,10 +6,14 @@ db.password=${db.password}
 neo4j.url=${neo4j.url}
 neo4j.database=${neo4j.database}
 
-citesphere_client_id=${citesphere.client.id}
-citesphere_client_secret=${citesphere.client.secret}
-citesphere_base_url=${citesphere.base.url}
+conceptpower_base_url=${conceptpower.base.url}
+conceptpower_id_url=/rest/Concept?id={concept_uri}
```
Comment (jdamerow) [inline quadriga/src/main/resources/config.properties]: that's the endpoint not the url

File: quadriga/src/main/resources/config.properties [hunk 28]
```diff
@@ -6,10 +6,14 @@ db.password=${db.password}
...
+
+citesphere_client_id=OAUTHCLIENT32
+citesphere_client_secret=03c2b326-88ba-4a38-866a-fc6f35ed0423
```
Comment (jdamerow) [inline quadriga/src/main/resources/config.properties]: these should not be in here (secret information!) and also should be defined in the pom.

Comment (jdamerow) [reply quadriga/src/main/resources/config.properties]: I regenerated the secret since this is compromised now.

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: quadriga/pom.xml [hunk 29]
```diff
@@ -44,7 +44,10 @@
         <citesphere.base.url></citesphere.base.url>
         <citesphere.client.id></citesphere.client.id>
         <citesphere.client.secret></citesphere.client.secret>
-
+        
+        <conceptpower.base.url></conceptpower.base.url>
+        <conceptpower.id.endpoint></conceptpower.id.endpoint>
```
Comment (jdamerow) [inline quadriga/pom.xml]: is that used anywhere?

**Review by jdamerow — CHANGES_REQUESTED** [review]
There are unaddressed comments.

File: quadriga/pom.xml (hunk 29 above)
Comment (pkharge) [reply quadriga/pom.xml]: @jdamerow , It's value is currently defined in config.properties. Shall I remove this entry from pom.xml or give the value here?

Comment (jdamerow) [reply quadriga/pom.xml]: remove it here

**Review by jdamerow — CHANGES_REQUESTED** [review]
there is a lot of code here that's also in the other PR I just reviewed, isn't there? Please reopen once the common code has been merged.

**Comment by jdamerow** [comment]
also, there are conflicts.

File: quadriga/src/main/resources/config.properties (hunk 15 above)
Comment (pooja-thalur) [reply quadriga/src/main/resources/config.properties:9]: removed the change from pom

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+    public void setConceptType(ConceptType conceptType) {
+        this.conceptType = conceptType;
+    }
+
+    @Override
+    public int compareTo(ConceptCache conceptCache) {
+        // If both old and new cache values are null/blank/empty, nothing has changed
+        // If old value is null/blank/empty, new value is not null/blank/empty, difference present
+        // If old value is not null/blank/empty, new value is null/blank/empty, difference present
+        // If both are not null/blank/empty, we need to check difference
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java]: this should probably go into the javadoc (which is missing)

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+        if(isDifferentList(conceptCache.getEqualTo(), this.getEqualTo())) return -1;
+        if(isDifferentList(conceptCache.getWordNetIds(), this.getWordNetIds())) return -1;
+        if (isDifferentString(conceptCache.getConceptList(), this.getConceptList())) return -1;
+        if (isDifferentString(conceptCache.getCreatorId(), this.getCreatorId())) return -1;
+        if (isDifferentString(conceptCache.getDescription(), this.getDescription())) return -1;
+        if (isDifferentString(conceptCache.getId(), this.getId())) return -1;
+        if (isDifferentString(conceptCache.getPos(), this.getPos())) return -1;
+        if (isDifferentString(conceptCache.getTypeId(), this.getTypeId())) return -1;
+        if (isDifferentString(conceptCache.getUri(), this.getUri())) return -1;
+        if (isDifferentString(conceptCache.getWord(), this.getWord())) return -1;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptCache.java]: this is not following our formatting guidelines. or it could be put as an OR statement (formatted one condition a line), so that only once -1 is returned.

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptType.java [hunk 32]
```diff
@@ -0,0 +1,81 @@
...
+        // If old value is null/blank, new value is not null/blank, difference present
+        // If old value is not null/blank, new value is null/blank, difference present
+        // If both are not null/blank, we need to check difference
+        
+        if (type == null ) return -1;
+        if(isDifferentString(type.getId(), this.getId())) return -1;
+        if(isDifferentString(type.getDescription(), this.getDescription())) return -1;
+        if(isDifferentString(type.getName(), this.getName())) return -1;
+        if(isDifferentString(type.getUri(), this.getUri())) return -1;
+        return 0;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/model/ConceptType.java:83]: same as before

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+    private RestTemplate restTemplate;
+
+    public ConceptPowerConnectorServiceImpl() {
+        restTemplate = new RestTemplate();
+    }
+
+    @Override
+    public ConceptPowerReply getConceptPowerReply(String conceptURI) {
+        Map<String, String> pathVariables = new HashMap<>();
+        pathVariables.put("concept_uri", conceptURI);
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java]: these are not path variables, are they? but parameters?

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+                logger.error("No response returned from ConceptPower for URI: " + conceptURI);
+                if (conceptPowerURL == null || conceptPowerURL.equals("")) {
+                    logger.error("ConceptPowerURL was found to be blank or null");
+                }
+            }
+        } catch (RestClientException e) {
+            logger.error("Could not get concept for URI: " + conceptURI + " at URL: " + conceptPowerURL, e);
+        }
+
+        return null;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerConnectorServiceImpl.java]: it looks like don't where you use the return value, you don't check for null, which means it'll throw a nullpointer if conceptpower is down? or throws an error? The better approach would probably be throw an exception if there is no response or an exception from conceptpower, and then handle it appropriate where it makes sense.

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 35]
```diff
@@ -0,0 +1,207 @@
...
+     * Updates Concept cache
+     * @param conceptCache
+     * @param conceptCacheOld
+     * @param uri
+     * @return
+     */
+    private boolean updateConceptCache(ConceptCache conceptCache, ConceptCache conceptCacheOld, String uri) {
+        
+        // ConceptPower returned a concept and either no conceptCache entry exists in the DB
+        // or if one exists, it is different from the current conceptCache entry
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: this should probably be part of the javadoc

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 36]
```diff
@@ -0,0 +1,207 @@
...
+        return false;
+    }
+
+    @Override
+    public ConceptCache mapConceptPowerReplyToConceptCache(ConceptPowerReply conceptPowerReply) {
+        // If we get multiple ConceptPower entries in the reply, we use the first one
+        List<ConceptEntry> conceptEntries = conceptPowerReply.getConceptEntries();
+        ConceptCache conceptCache = null;
+
+        if (conceptEntries != null && !conceptEntries.isEmpty()) {
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: turn this if around and just return if this condition is not met. It'll unindent all the following code.

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/service/impl/MappedTripleGroupServiceImpl.java [hunk 37]
```diff
@@ -145,7 +151,13 @@ public MappedTripleGroup get(String collectionId, MappedTripleType mappedTripleT
         // we create a new MappedTripleGroup entry
         if (mappedTripleGroup == null) {
             mappedTripleGroup = add(collectionId, mappedTripleType);
+            
+            // If mappedTripleGroup add fails
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/service/impl/MappedTripleGroupServiceImpl.java:156]: shouldn't this be made known to the upper layers via an exception?

File: quadriga/src/main/resources/config.properties (hunk 15 above)
Comment (jdamerow) [reply quadriga/src/main/resources/config.properties:9]: there is no `conceptpower.base.url` property in the pom, is there? If this property is required here, it needs to be added to the pom.

File: quadriga/src/main/resources/config.properties [hunk 38]
```diff
@@ -6,10 +6,14 @@ db.password=${db.password}
 neo4j.url=${neo4j.url}
 neo4j.database=${neo4j.database}
 
+conceptpower_base_url=${conceptpower.base.url}
+conceptpower_id_endpoint=/rest/Concept?id={concept_uri}
+
 citesphere_client_id=${citesphere.client.id}
-citesphere_client_secret=${citesphere.client.secret}
+citesphere_client_secret=${citesphere.client.id}
```
Comment (jdamerow) [inline quadriga/src/main/resources/config.properties]: this does not look right

**Review by jdamerow — CHANGES_REQUESTED** [review]
It really throws me off that the ConceptPower classes capitalize the "P" 😅 the app is called "Conceptpower". Maybe change the class names? :)

File: quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java [hunk 39]
```diff
@@ -0,0 +1,209 @@
...
+    
+    @Value("${conceptCacheUpdateInterval}")
+    private Integer conceptCacheUpdateInterval;
+
+    @Override
+    public ConceptCache getConceptByUri(String uri) throws ConceptpowerNoResponseException {
+
+        ConceptCache conceptCache = conceptCacheService.getConceptByUri(uri);
+        
+        conceptCacheUpdateInterval=0;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/core/conceptpower/service/impl/ConceptPowerServiceImpl.java]: this seems wrong

File: quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java [hunk 40]
```diff
@@ -22,34 +49,55 @@ public class DisplayCollectionController {
...
             logger.error("Couldn't find collection ", e);
             return "error404Page";
         }
-        
+
+        model.addAttribute("collection", collection);
+
+//      Determine page number and size for network pagination      
+        page = (page == null || page < 0) ? 0 : page - 1;
+        size = (size == null || size < 1) ? 10 : size;
```
Comment (jdamerow) [inline quadriga/src/main/java/edu/asu/diging/quadriga/web/DisplayCollectionController.java]: instead of checking for null you can just specify a default value in the RequestParam annotation

File: quadriga/src/main/resources/config.properties (hunk 15 above)
Comment (jdamerow) [reply quadriga/src/main/resources/config.properties:9]: I think this has still not been added to the pom?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Tests are failing
